### DISCUSSION TITLE
How to reuse custom styles

### DISCUSSION THREAD

**Discussion by 7iomka** [mantinedev/mantine#2812] [Q&A]
Hi.
I configured once my segmented control like this.

```tsx

<SegmentedControl
  radius="lg"
  size="lg"
  styles={(
    theme,
    { color, shouldAnimate, transitionTimingFunction, transitionDuration },
  ) => {
    const colors = theme.fn.variant({ variant: 'filled', color });
    const activeColor =
      color in theme.colors
        ? colors.background
        : theme.colorScheme === 'dark'
        ? theme.colors.dark[5]
        : theme.white;
    return {
      root: {
        padding: 0,
        background: 'none',
        overflow: 'visible',
        marginRight: -10,
        marginBottom: -10,
        flexWrap: 'wrap',
      },
      control: {
        marginRight: 10,
        marginBottom: 10,

        color:
          theme.colorScheme === 'dark' ? theme.colors.dark[1] : theme.colors.gray[7],
        transition: `color ${shouldAnimate ? 0 : transitionDuration}ms ${
          transitionTimingFunction || theme.transitionTimingFunction
        }`,

        '&:not(:first-of-type)': {
          border: 0,
        },

        '&:hover': {
          color: theme.colorScheme === 'dark' ? theme.colors.dark[0] : theme.black,
        },
      },
      label: {
        border: `1px solid ${theme.colors.gray[7]}`,
        height: '100%',
      },

      labelActive: {
        cursor: 'default',
        backgroundColor: activeColor,
        borderColor: activeColor,
      },
      active: {
        display: 'none',
      },
    };
  }}
  data={[]}
/>
```

And I want to reuse this styles to prevent copy-paste the same code multiple times.
Also I want to keep availability for customization through styles if I need a little different behavior.

I tried to extract code from styles inside separate file

```ts
import type { MantineNumberSize, SegmentedControlStylesParams } from '@mantine/core';
import { createStyles } from '@mantine/core';

export interface SegmentedCardsStylesParams extends SegmentedControlStylesParams {
  gap: MantineNumberSize;
}

export const useSegmentedCardsStyles = createStyles(
  (
    theme,
    {
      gap,
      color,
      shouldAnimate,
      transitionTimingFunction,
      transitionDuration,
    }: SegmentedCardsStylesParams,
  ) => {
    const gapValue = theme.fn.size({
      size: gap,
      sizes: theme.spacing,
    });
    const colors = theme.fn.variant({ variant: 'filled', color });
    const activeColor =
      color in theme.colors
        ? colors.background
        : theme.colorScheme === 'dark'
        ? theme.colors.dark[5]
        : theme.white;

    return {
      root: {
        padding: 0,
        background: 'none',
        overflow: 'visible',
        marginRight: gapValue * -1,
        marginBottom: gapValue * -1,
        flexWrap: 'wrap',
      },
      control: {
        marginRight: gapValue,
        marginBottom: gapValue,

        color: theme.colorScheme === 'dark' ? theme.colors.dark[1] : theme.colors.gray[7],
        transition: `color ${shouldAnimate ? 0 : transitionDuration}ms ${
          transitionTimingFunction || theme.transitionTimingFunction
        }`,

        '&:not(:first-of-type)': {
          border: 0,
        },

        '&:hover': {
          color: theme.colorScheme === 'dark' ? theme.colors.dark[0] : theme.black,
        },
      },
      label: {
        border: `1px solid ${theme.colors.gray[7]}`,
        height: '100%',
      },

      labelActive: {
        cursor: 'default',
        backgroundColor: activeColor,
        borderColor: activeColor,
      },
      active: {
        display: 'none',
      },
    };
  },
);

```

But now we have a problem when we try to use this hook
```tsx
const { classes: optionsClasses } = useSegmentedCardsStyles({
  gap: 10,
});
```

We have to pass all props to our hook, which is required by `SegmentedControlStylesParams`, so how to figure out?
When we just used inline styles approach, it used default props. How I can use this default props to pass to hook?
Help me please.
Also note that I don't want to declare in my custom hook only my own params, how documentation says through examples, but I want to use the default props.

Thanks for any help.

**Comment by rtivital** [comment]
try `export interface SegmentedCardsStylesParams extends Partial<SegmentedControlStylesParams>`

**Reply by 7iomka** [reply]
> export interface SegmentedCardsStylesParams extends Partial<SegmentedControlStylesParams>

It looks like you didn't read my post to the end :(
I wrote that I want to use access to default props in my hook. If I use Partial in type declaration, then inside the hook I will have errors -  because I want to use props which is default settled internally but type says that it can be undefined.

**Reply by 7iomka** [reply]
@rtivital Also, in my hook I use 
```
const colors = theme.fn.variant({ variant: 'filled', color });
 const activeColor =
      color in theme.colors
        ? colors.background
        : theme.colorScheme === 'dark'
        ? theme.colors.dark[5]
        : theme.white;
```

So, even if I change my type declaration, the final result of hook usage wiil be wrong, because I can't pass color when I call the hook
```
const { classes: optionsClasses } = useSegmentedCardsStyles({
  gap: 10,
});
```

Because of this I can't get the same result as I get when I used inline styles
So, I need to pass somehow default props into hook.
Help me please.

**Reply by rtivital** [reply]
Sorry, I do not really get what "default props" is in this context, can you provide a minimal example showing what you want to achieve?

**Comment by 7iomka** [comment]
@rtivital I mean this default props used under hood, when you just use inline styles, you can override some of it, but in rest will be used the one from provider config or defaultProps.
https://github.com/mantinedev/mantine/blob/master/src/mantine-core/src/SegmentedControl/SegmentedControl.tsx#L102
But when I have to create a separate hook like in code samples from above, I don't have access to this props values, so, event if I change typings, in my hook I don't pass this props which SegmentedControl uses under hood.
So problem is to have access to it into my hook to operate with them.

**Reply by rtivital** [reply]
I still do not get the issue. Why you would not just create a wrapper component that will manage these props, in this component you will be able to call `useComponentDefaultProps` and get default props for SegmentedControl component.

**Reply by 7iomka** [reply]
> I still do not get the issue. Why you would not just create a wrapper component that will manage these props, in this component you will be able to call `useComponentDefaultProps` and get default props for SegmentedControl component.

I need some styles for SegmentedControl, but I don't want to repeat whole stuff in inline styles.
When you put some config in inline styles, you already have acces to params setted from provider/internally.
I want to keep this simple way to configure through createStyles, but now when I'm extractes setup into this hook, I need all of this 'params' to get from somewhere, and to pass in hook usage, like you in source code of this component.
useComponentDefaultProps required 2nd and 3rd arguments, so I need to understand how to use my hook multiple times in different files, easily.
You understand the problem?

**Reply by rtivital** [reply]
> I need some styles for SegmentedControl, but I don't want to repeat whole stuff in inline styles.

What is wrong with creating a wrapper component for that?

> When you put some config in inline styles, you already have acces to params setted from provider/internally.

Why not store inline styles in a variable? 

> I need all of this 'params' to get from somewhere, and to pass in hook usage, like you in source code of this component.

Create a wrapper component and provide any params there

> useComponentDefaultProps required 2nd and 3rd arguments, so I need to understand how to use my hook multiple times in different files, easily.

Create a wrapper hook that will include first and second argument

> You understand the problem?

Still, no

**Reply by 7iomka** [reply]
@rtivital 
I created a demo for my case
https://codesandbox.io/s/stupefied-hertz-5slhbi?file=/src/App.tsx

When I extract into hook all repeatable code I want to keep the same behavior as with inline styles
In the demo I used extracted in createStyles hook styles inside a custom component which uses SegmentedControl
But, because I don't passed all props settled in provider (or in source code as defaultProps) but I try to use it in my hook, I have broken styles and error in hook usage. At the moment I don't understand from where I can take this props?
> Why not store inline styles in a variable?

Because this adds complexity in case I need to use additional styles in some variant of the component.

**Reply by 7iomka** [reply]
@rtivital No ideas? :(

**Reply by rtivital** [reply]
Sorry, I still do not understand the issue. My main question is: why haven't you created a wrapper component for SegmentedControl that will manage all props and styles?

**Reply by 7iomka** [reply]
> Sorry, I still do not understand the issue. My main question is: why haven't you created a wrapper component for SegmentedControl that will manage all props and styles?

I want to create handy reusable hooks that I can give to end users so they can use them along with other hooks or styles if they wish with some of mantine components directly.
Instead you propose to make a component wrapper for it. 
Why even come up with the idea of hooks then?

**Reply by rtivital** [reply]
Yes, it will be much easier to ship components with styles, this is an idea behind components libraries, for example, Mantine provides components, not hooks to style other components.
